File: src/Total.Recall.Scanners.TypeScript/src/coverage.ts
/**
 * Cobertura coverage parser — matches the Python / .NET scanner shape.
 *
 * Tolerant of missing / malformed XML (returns []). When both <lines>
 * and <methods> are present, <lines> is the source of truth for class
 * totals (matches coverage.py / istanbul output); <methods> only
 * contributes the per-symbol uncovered-line breakdown.
 */

import { readFileSync } from "node:fs";
import { XMLParser } from "fast-xml-parser";

const SCHEMA_VERSION = 1;

export interface CoverageRecord {
  schemaVersion: number;
  className: string;
  linesCovered: number;
  linesTotal: number;
  coveragePercent: number;
  uncoveredMethods: { methodName: string; uncoveredLines: number[]; totalLines: number }[];
  existingTests: null;
  testabilityScore: null;
}

export function parseCobertura(xmlPath: string): CoverageRecord[] {
  let raw: string;
  try {
    raw = readFileSync(xmlPath, "utf8");
  } catch {
    return [];
  }
  let parsed: any;
  try {
    const parser = new XMLParser({
      ignoreAttributes: false,
      attributeNamePrefix: "@_",
      allowBooleanAttributes: true,
      isArray: (name) => ["package", "class", "method", "line"].includes(name),
    });
    parsed = parser.parse(raw);
  } catch {
    return [];
  }

  const packages = parsed?.coverage?.packages?.package ?? [];
  const records: CoverageRecord[] = [];
  for (const pkg of packages) {
    const classes = pkg?.classes?.class ?? [];
    for (const cls of classes) {
      records.push(classRecord(cls));
    }
  }
  records.sort((a, b) => a.className.localeCompare(b.className));
  return records;
}

function classRecord(cls: any): CoverageRecord {
  const className = cls?.["@_name"] ?? "";
  let linesTotal = 0;
  let linesCovered = 0;
  const uncoveredMethods: CoverageRecord["uncoveredMethods"] = [];

  const linesNode = cls?.lines?.line;
  if (Array.isArray(linesNode)) {
    for (const line of linesNode) {
      linesTotal++;
      const hits = Number(line?.["@_hits"] ?? 0);
      if (hits > 0) linesCovered++;
    }
  }

  const methodsNode = cls?.methods?.method;
  if (Array.isArray(methodsNode)) {
    for (const method of methodsNode) {
      const methodName = method?.["@_name"] ?? "";
      const methodLines = method?.lines?.line ?? [];
      const lines = Array.isArray(methodLines) ? methodLines : [methodLines].filter(Boolean);
      const uncovered: number[] = [];
      let total = 0;
      for (const line of lines) {
        total++;
        const hits = Number(line?.["@_hits"] ?? 0);
        if (hits === 0) uncovered.push(Number(line?.["@_number"] ?? 0));
      }
      if (methodName) {
        uncoveredMethods.push({ methodName, uncoveredLines: uncovered, totalLines: total });
        if (!Array.isArray(linesNode)) {
          linesTotal += total;
          linesCovered += total - uncovered.length;
        }
      }
    }
  }

  const coveragePercent =
    linesTotal === 0 ? 0 : Math.round((linesCovered / linesTotal) * 10000) / 100;

  return {
    schemaVersion: SCHEMA_VERSION,
    className,
    linesCovered,
    linesTotal,
    coveragePercent,
    uncoveredMethods,
    existingTests: null,
    testabilityScore: null,
  };
}

File: src/Total.Recall.Scanners.Python/src/total_recall_scan/coverage.py
"""Cobertura XML -> coverage-gaps.jsonl converter.

The coverage.py tool emits Cobertura via ``coverage xml``. We parse the
``<class>`` elements (which represent Python modules in coverage.py's
output) and collapse line-level hit counts into a per-class summary plus
``uncoveredMethods`` listing methods whose every line is uncovered.

Schema documented in docs/SCANNER_SCHEMA.md section 2.
"""

from __future__ import annotations

import xml.etree.ElementTree as ET
from pathlib import Path

from . import __schema_version__


def parse_cobertura(xml_path: str | Path) -> list[dict]:
    """Parse a Cobertura XML file and return CoverageGap-shaped dicts.

    Returns an empty list if the file is missing or unparseable — scanners
    should never abort the whole scan on a malformed coverage report.
    """
    path = Path(xml_path)
    if not path.exists():
        return []

    try:
        tree = ET.parse(path)
    except ET.ParseError:
        return []

    root = tree.getroot()
    records: list[dict] = []

    for cls in root.iter("class"):
        record = _record_for_class(cls)
        if record is not None:
            records.append(record)

    records.sort(key=lambda r: r["className"])
    return records


def _record_for_class(cls: ET.Element) -> dict | None:
    name = cls.attrib.get("name") or ""
    filename = cls.attrib.get("filename") or ""
    if not name:
        return None

    lines_total = 0
    lines_covered = 0
    method_uncovered: dict[str, list[int]] = {}
    method_total: dict[str, int] = {}

    # Prefer the class-level <lines> block for totals (it is the source of
    # truth in coverage.py output and includes lines that don't belong to a
    # method). Use <methods> only to attribute uncovered lines to symbols.
    lines_node = cls.find("lines")
    if lines_node is not None:
        for line in lines_node.findall("line"):
            lines_total += 1
            hits = int(line.attrib.get("hits", "0"))
            if hits > 0:
                lines_covered += 1

    methods_node = cls.find("methods")
    if methods_node is not None:
        for method in methods_node.findall("method"):
            method_name = method.attrib.get("name") or ""
            uncovered: list[int] = []
            total = 0
            for line in method.iter("line"):
                total += 1
                hits = int(line.attrib.get("hits", "0"))
                if hits == 0:
                    uncovered.append(int(line.attrib.get("number", "0")))
            if method_name:
                method_uncovered[method_name] = uncovered
                method_total[method_name] = total
                # Fall back to method-level totals only if <lines> was absent.
                if lines_node is None:
                    lines_total += total
                    lines_covered += total - len(uncovered)

    coverage_pct = (
        round((lines_covered / lines_total) * 100.0, 2) if lines_total > 0 else 0.0
    )

    uncovered_methods = [
        {
            "name": method_name,
            "signature": method_name,
            "uncoveredLines": uncovered_lines,
            "totalLines": method_total[method_name],
        }
        for method_name, uncovered_lines in method_uncovered.items()
        if uncovered_lines
    ]

    return {
        "schemaVersion": __schema_version__,
        "className": name,
        "filePath": filename.replace("\\", "/") if filename else "",
        "linesCovered": lines_covered,
        "linesTotal": lines_total,
        "coveragePercent": coverage_pct,
        "uncoveredMethods": uncovered_methods,
        "existingTests": None,
        "testabilityScore": None,
    }
